src/components/TodoList.tsx

```diff
@@ -0,0 +1,102 @@
+import React from 'react';
+import { Todo, TodoItem } from './TodoItem';
+
+interface TodoListProps {
+  todos: Todo[];
+  filter: string;
+  editingTodoId: number | null;
+  newTitle: string;
+  inputRef: React.RefObject<HTMLInputElement>;
+  loading: boolean;
+  onToggleTodo: (id: number) => void;
+  onEditTodo: (todo: Todo) => void;
+  onDeleteTodo: (id: number) => void;
+  onSaveTitle: (id: number) => void;
+  onChangeNewTitle: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  deletingTodoIds: number[];
+  tempTodo?: Todo | null;
+  updatingTodoId: number | null;
+  batchUpdatingIds: number[];
+}
+
+export const TodoList: React.FC<TodoListProps> = ({
+  todos,
+  filter,
+  editingTodoId,
+  newTitle,
+  inputRef,
+  onToggleTodo,
+  onEditTodo,
+  onDeleteTodo,
+  onSaveTitle,
+  onChangeNewTitle,
+  deletingTodoIds,
+  tempTodo,
+  updatingTodoId,
+  batchUpdatingIds,
```

Comment:
There are a lot of props which are doing almost the same (for example edit todo). Please use less of them